qiskit_experiments/library/randomized_benchmarking/rb_experiment.py

```diff
@@ -133,86 +141,128 @@ def circuits(self) -> List[QuantumCircuit]:
+        # Add metadata for each circuit
+        for circ, seq in zip(circuits, sequences):
+            circ.metadata = {
+                "experiment_type": self._type,
+                "xval": len(seq),
+                "group": "Clifford",
+                "physical_qubits": self.physical_qubits,
+            }
         return circuits
 
-    def _sample_circuits(self, lengths: Iterable[int], rng: Generator) -> List[QuantumCircuit]:
-        """Return a list RB circuits for the given lengths.
-
-        Args:
-            lengths: A list of RB sequences lengths.
-            seed: Seed or generator object for random number
-                  generation. If None default_rng will be used.
+    def _sample_sequences(self) -> List[Sequence[SequenceElementType]]:
+        """Sample RB sequences
 
         Returns:
-            A list of :class:`QuantumCircuit`.
+            A list of RB sequences.
         """
-        circuits = []
-        for length in lengths if self._full_sampling else [lengths[-1]]:
-            elements = self._clifford_utils.random_clifford_circuits(self.num_qubits, length, rng)
-            element_lengths = [len(elements)] if self._full_sampling else lengths
-            circuits += self._generate_circuit(elements, element_lengths)
-        return circuits
-
-    def _generate_circuit(
-        self, elements: Iterable[Clifford], lengths: Iterable[int]
+        rng = default_rng(seed=self.experiment_options.seed)
+        sequences = []
+        if self._full_sampling:
+            for _ in range(self.experiment_options.num_samples):
+                for length in self.experiment_options.lengths:
+                    sequences.append(self._sample_sequence(length, rng))
+        else:
+            for _ in range(self.experiment_options.num_samples):
+                longest_seq = self._sample_sequence(max(self.experiment_options.lengths), rng)
+                for length in self.experiment_options.lengths:
+                    sequences.append(longest_seq[:length])
+
+        return sequences
+
+    def _sequences_to_circuits(
+        self, sequences: List[Sequence[SequenceElementType]]
     ) -> List[QuantumCircuit]:
-        """Return the RB circuits constructed from the given element list.
-
-        Args:
-            elements: A list of Clifford elements
-            lengths: A list of RB sequences lengths.
+        """Convert a RB sequence into circuit and append the inverse to the end.
 
         Returns:
-            A list of :class:`QuantumCircuit`s.
-
-        Additional information:
-            The circuits are constructed iteratively; each circuit is obtained
-            by extending the previous circuit (without the inversion and measurement gates)
+            A RB circuit.
         """
-        qubits = list(range(self.num_qubits))
         circuits = []
+        for i, seq in enumerate(sequences):
+            if self._full_sampling or i % len(self.experiment_options.lengths) == 0:
```

Comment:
Another option would be calling this per sample (seed), rather than generating a big single sequence of all seeds. This makes logic simpler and allows us to use multi threading in future updates.

Reply:
Let's leave it as an option for future. Note for the future:
```
        def split_list(org_list, num_sublist):
            for idx in range(0, len(org_list), num_sublist):
                yield org_list[idx:idx + num_sublist]

        circuits = []
        for sequence in split_list(sequences, len(self.experiment_options.lengths)):
            prev_elem, prev_seq = self.__identity_clifford(), []
            for seq in sequence:
                if self._full_sampling:
                    prev_elem, prev_seq = self.__identity_clifford(), []
                    ...
```